x/shield/keeper/proposal.go

```diff
@@ -17,3 +24,415 @@ func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID, poolID uint64, loss sdk
+		uncoveredPurchase = purchased.Sub(provider.DelegationBonded)
+	}
+	payoutFromUnbonding := payout.Sub(payoutFromDelegation)
+
+	if payoutFromDelegation.IsPositive() {
+		k.PayFromDelegation(ctx, provider.Address, payoutFromDelegation)
+	}
+
+	if payoutFromUnbonding.IsZero() {
+		return nil
+	}
+
+	unbondingDelegations := k.GetSortedUnbondingDelegations(ctx, provider.Address)
+	for _, ubd := range unbondingDelegations {
+		if !payoutFromUnbonding.IsPositive() {
+			break
+		}
+		entry := ubd.Entries[0]
+
+		// If purchased is not fully covered, cover purchased first.
+		remainingUbd := sdk.MaxInt(entry.Balance.Sub(uncoveredPurchase), sdk.ZeroInt())
+		uncoveredPurchase = sdk.MaxInt(uncoveredPurchase.Sub(entry.Balance), sdk.ZeroInt())
+		if remainingUbd.IsZero() {
+			continue
+		}
+
+		// Make payout after purchased is fully covered.
+		payoutFromThisUbd := sdk.MinInt(payoutFromUnbonding, remainingUbd)
+		k.PayFromUnbonding(ctx, ubd, payoutFromThisUbd)
+
+		payoutFromUnbonding = payoutFromUnbonding.Sub(payoutFromThisUbd)
+	}
+
+	return nil
+}
+
+// PayFromDelegation reduce provider's delegations and transfer tokens to the shield module account.
+func (k Keeper) PayFromDelegation(ctx sdk.Context, delAddr sdk.AccAddress, payout sdk.Int) {
+	provider, found := k.GetProvider(ctx, delAddr)
+	if !found {
+		panic(types.ErrProviderNotFound)
+	}
+	totalDelAmount := provider.DelegationBonded
+
+	delegations := k.sk.GetAllDelegatorDelegations(ctx, delAddr)
+	payoutRatio := payout.ToDec().Quo(totalDelAmount.ToDec())
+	remaining := payout
+	for i := range delegations {
+		if !remaining.IsPositive() {
+			return
+		}
+
+		val, found := k.sk.GetValidator(ctx, delegations[i].GetValidatorAddr())
+		if !found {
+			panic("validator is not found")
+		}
+		delAmount := val.TokensFromShares(delegations[i].GetShares()).TruncateInt()
+		var ubdAmount sdk.Int
+		if i == len(delegations)-1 {
+			ubdAmount = remaining
+		} else {
+			ubdAmount = payoutRatio.MulInt(delAmount).TruncateInt()
+			if ubdAmount.LT(remaining) && ubdAmount.LT(delAmount) {
+				ubdAmount = ubdAmount.Add(sdk.OneInt())
+			}
+			remaining = remaining.Sub(ubdAmount)
+		}
+		ubdShares, err := val.SharesFromTokens(ubdAmount)
+		if err != nil {
+			panic(err)
+		}
+		k.UndelegateShares(ctx, delegations[i].DelegatorAddress, delegations[i].ValidatorAddress, ubdShares)
+	}
+}
+
+// PayFromUnbonding reduce provider's unbonding delegation and transfer tokens to the shield module account.
+func (k Keeper) PayFromUnbonding(ctx sdk.Context, ubd staking.UnbondingDelegation, payout sdk.Int) {
+	unbonding, found := k.sk.GetUnbondingDelegation(ctx, ubd.DelegatorAddress, ubd.ValidatorAddress)
+	if !found {
+		panic("unbonding delegation is not found")
+	}
+
+	// Update unbonding delegations between the delegator and the validator.
+	for i := range unbonding.Entries {
+		if unbonding.Entries[i].Balance.Equal(ubd.Entries[0].Balance) && unbonding.Entries[i].CompletionTime.Equal(ubd.Entries[0].CompletionTime) {
```

Comment:
what is this doing? why not `if i == 0`?

Reply:
`ubd` comes from `GetSortedUnbondingDelegations()` so it only has one entry, while `unbonding` comes from `GetUnbondingDelegation` and has multiple entries. Their indices are thus different.